contracts/Eip712Forwrader.sol

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier:MIT
+pragma solidity ^0.6.2;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/cryptography/ECDSA.sol";
+import "./utils/GsnUtils.sol";
+
+// a Generic EIP712 forwarder.
+// actual struct has to START with known fields, but may contain other fields
+contract Eip712Forwarder {
+    using ECDSA for bytes32;
+
+    //all valid requests must start with this prefix.
+    // (note that the prefix is not a valid type by itself: it doesn't end with ")" (or ",", if more fields are added)
+    string constant typehashPrefix = "Forward(address sender, address target, bytes encodedFunction, uint256 gasLimit, address forwarder";
```

Comment:
Can we extract it to be a struct? So instead of 'gasData' and 'relayData' we have 'ForwarderData' and 'GSNData'?

Reply:
Also, missing nonce, right?

Reply:
And, use original Ethereum names. Together, this structure is 
```from to data gas gasPrice nonce```, only missing ```value```

Also, we can include 'value' in this forwarder. If the paymaster sends ether here, it should accept it and make it available for 'verifyAndCall', what do you think?

Reply:
Also, we could close that bracket if we make it a suffix, not a prefix (requires it to be last alphabetically)
```RelayRequest(ForwarderData fd, GSNData gd)GSNData(pctRelayFee,baseRelayFee)ZZZForwarderData(from,to,data,gas,gasPrice,nonce)```

Reply:
1. its deliberately prefix, since these are supposed to be the most important fields. putting it last, means you can add as much garbage at the beginning.
2. making it a struct is problematic: the parsing will become much more complex. the prefix is "RelayData data" and then the "RelayData" is defined somewhere - in a sorted list of types...
AAATypename and ZZZTypename are not good either...

I want to use the original names, I just dislike "data" as it seems to me more of an anachronism than intended name. "method" or "function" is better.